src/libraries/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Windows/X509Pal.PublicKey.cs

```diff
@@ -53,10 +74,21 @@ public AsymmetricAlgorithm DecodePublicKey(Oid oid, byte[] encodedKeyValue, byte
             }
         }
 
-        private static ECDsa DecodeECDsaPublicKey(CertificatePal certificatePal)
+        private static TAlgorithm DecodeECPublicKey<TAlgorithm>(
+            CertificatePal certificatePal,
+            Func<CngKey, TAlgorithm> factory,
+            Action<TAlgorithm, ECParameters> import)
+                where TAlgorithm : AsymmetricAlgorithm, new()
         {
-            ECDsa ecdsa;
-            using (SafeBCryptKeyHandle bCryptKeyHandle = ImportPublicKeyInfo(certificatePal.CertContext))
+            TAlgorithm key;
+            CryptImportPublicKeyInfoFlags importFlags = CryptImportPublicKeyInfoFlags.NONE;
+
+            if (typeof(ECDiffieHellmanCng) == typeof(TAlgorithm))
```

Comment:
We're doing strict equality on the type parameter, which means we have some expectations as to what's going on.

Is it worth adding a debugging aid, like

```C#
Debug.Assert(typeof(TAlgorithm).BaseType?.BaseType == typeof(AsymmetricAlgorithm), $"TAlgorithm ({typeof(TAlgorithm).FullName}) should be an implementation type, not an algorithm type.");
```

Sure, the assert would be wrong if we ever subclassed ECDiffieHellmanCng, but it turns out, this if would be wrong, too :smile:.

Reply:
I think I'd rather take a bool to this method and let the caller decide if this flag is applicable instead of trying to determine from the `TAlgorithm`. Something like `bool restrictPublicKeyToEncryption` on this method. The ECDSA paths will specify false and the ECDH path will say true. Would that be better?

Reply:
Either a bool, or (since it's already a Windows-PAL-only member) just the CryptImportPublicKeyInfoFlags themselves (perhaps defaulted to none) seems fine to me.